pkg/globalnet/controllers/ingress_endpoints_controller.go

```diff
@@ -0,0 +1,203 @@
+	endpoints := from.(*corev1.Endpoints)
+
+	switch op {
+	case syncer.Create, syncer.Update:
+		return c.onCreateOrUpdate(endpoints, op)
+	case syncer.Delete:
+		return c.onDelete(endpoints)
+	}
+
+	return nil, false
+}
+
+func (c *ingressEndpointsController) onCreateOrUpdate(endpoints *corev1.Endpoints, op syncer.Operation) (runtime.Object, bool) {
+	key, _ := cache.MetaNamespaceKeyFunc(endpoints)
+
+	klog.Infof("%q ingress Endpoints %s for service %s", op, key, c.svcName)
+
+	// Get list of current endpoint IPs
+	epIPs := map[string]bool{}
+
+	for _, subset := range endpoints.Subsets {
+		for _, addr := range subset.Addresses {
+			epIPs[addr.IP] = true
+		}
+	}
+
+	// Get List of existing ingressIPs
+	selector := labels.SelectorFromSet(map[string]string{ServiceRefLabel: endpoints.Name}).String()
+
+	existingIngressIPs, err := c.ingressIPs.Namespace(endpoints.Namespace).List(context.TODO(), metav1.ListOptions{LabelSelector: selector})
+	if err != nil {
+		return nil, true
+	}
+
+	// Delete unused ingressIPs
+	for _, ingressIP := range existingIngressIPs.Items {
+		annotations := ingressIP.GetAnnotations()
+		if epIP, ok := annotations[headlessSvcEndpointsIP]; ok {
+			if _, used := epIPs[epIP]; used {
+				// ingressIP is still used
+				continue
+			}
+		}
+
+		err := c.ingressIPs.Namespace(ingressIP.GetNamespace()).Delete(context.TODO(), ingressIP.GetName(), metav1.DeleteOptions{})
+		if err != nil {
+			return nil, true
+		}
+	}
+
+	// Create new ingressIPs
+	for epIP := range epIPs {
+		ingressIP := newIngressIP(endpoints.Name, endpoints.Namespace, epIP)
+
+		ingressIP.ObjectMeta.Annotations = map[string]string{
+			headlessSvcEndpointsIP: epIP,
+		}
+
+		ingressIP.Spec = submarinerv1.GlobalIngressIPSpec{
+			Target:     submarinerv1.HeadlessServiceEndpoints,
+			ServiceRef: &corev1.LocalObjectReference{Name: c.svcName},
+		}
+
+		uIngressIP, _, err := util.ToUnstructuredResource(ingressIP, c.config.RestMapper)
+		if err != nil {
+			return nil, true
+		}
+
+		_, err = c.ingressIPs.Namespace(ingressIP.GetNamespace()).Create(context.TODO(), uIngressIP, metav1.CreateOptions{})
+		if err != nil {
+			return nil, true
+		}
+	}
+
+	return nil, false
+}
+
+func (c *ingressEndpointsController) onDelete(endpoints *corev1.Endpoints) (runtime.Object, bool) {
+	key, _ := cache.MetaNamespaceKeyFunc(endpoints)
+
+	selector := labels.SelectorFromSet(map[string]string{ServiceRefLabel: endpoints.Name}).String()
+
+	err := c.ingressIPs.Namespace(endpoints.Namespace).DeleteCollection(context.TODO(), metav1.DeleteOptions{},
+		metav1.ListOptions{LabelSelector: selector})
+	if err != nil {
+		return nil, true
+	}
+
+	klog.Infof("Ingress Endpoints %s for service %s deleted", key, c.svcName)
+
+	return nil, false
+}
+
+func newIngressIP(name, namespace, epIP string) *submarinerv1.GlobalIngressIP {
+	return &submarinerv1.GlobalIngressIP{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("ep-%.44s-%.15s", name, epIP),
```

Comment:
So if the user creates two different endpoints for different services where the initial 44 chars of names match and epIP matches, we would have a problem here. Is this correct? Its a very corner case, but just want to hear from you if I'm reading the code correctly.

Reply:
Agree, using a hast makes it difficult to recognise it. Since its a very corner case, lets continue this way.